proposals/0000-extension-lifecycle-framework.md

```diff
@@ -0,0 +1,108 @@
+
+## 2. Proposed Change Specification
+
+### 2.1. Categories and Lifecycle
+Language extensions will be classified into the following categories
+ * `Experimental` extensions are undergoing active development. The syntax and semantics that are enabled by the extension are likely to change regularly. It is expected that most new language extensions will begin as experimental. At the time of writing, `LinearTypes` and `OverloadedRecordUpdate` seem to be in this category.
+ 
+ * `Mature` extensions are considered to be finished and are not expected to undergo regular changes. These features can be used without worry of unexpected changes, and they are not known to contain serious design or implementation deficiencies. Any breaking change to a mature extension will be announced well in advance of the change being made, with a migration path provided if possible. Some extensions in this category might be `MultiWayIf`, `MonoLocalBinds`, and `ViewPatterns`.
+
+ * `Deprecated` extensions are considered to be design or implementation dead ends, and should probably not be used in new code. Deprecating an extension _must_ come with a statement about the longevity of the extension. This need not be a commitment to remove it at a concrete time, but is a commitment to either remove it or transition it to another state at a time stated by GHC HQ when it is marked as `Deprecated`. Any extension will be deprecated prior to removal. Deprecated extensions are expected to include `IncoherentInstances`, `OverlappingInstances`, `Rank2Types`, `RecordPuns`, and `TypeInType`.
+ 
+ * `Legacy` extensions are explicitly not recommended for new code, much like `Deprecated`, however, they are expected to be supported indefinitely. This may be to maintain an older language set, maintain a bridge to a newer feature, or otherwise keep backward compatibility. Expected legacy extensions include `CUSKs`, `DeepSubsumption`,  and `NPlusKPatterns`.
+
+Above we list expected categorizations of several extensions based on statements in the user's guide. We do not attempt to prescribe any particular categorizations as part of this proposal and simply provide a small number of expectations from the perspective of someone reading the existing documentation.
+
+The expected extension lifecycle includes the following transitions:
+ * `Experimental` -> `Mature`
+ * `Experimental` -> `Deprecated`
+ * `Mature` -> `Legacy`
+ * `Mature` -> `Deprecated`
+ * `Legacy` -> `Deprecated`
+ * `Deprecated` -> `Legacy`
+
+However, it also seems plausible that new knowledge might from time to time cause a mature extension to once again be considered experimental, e.g. in the face of soundness bugs or subtle interactions with other features. We also do not rule out any transition explicity to allow for unforeseen circumstances. 
+
+For existing, or future, language sets such as `GHC2021` or `Haskell98`, it is expected that none of the contained extensions would be `Experimental`.
```

Comment:
Haskell 98 certainly includes `NPlusKPatterns`, which I think may end up in one of those categories. And over time, it doesn't seem implausible that we'd declare legacy an extension that had been in one of these default sets (after a very, very long notice period, of course)

Reply:
Right, things that are in a set can become deprecated or legacy, but when we create a set we probably shouldn't include anything in those states? e.g. GHC2021 does _not_ contain `NPlusKPatterns`

Reply:
It certainly seems like a good idea to as a matter of policy say that a new language set does not contain legacy or deprecated extensions. However, if a language set should be added after the fact for some reason, then perhaps it would include a legacy extension. Maybe resolving if a policy should exist for this and what exactly that would be is worthy of a separate discussion?

Reply:
I'm loathe to impose policy here. I think that the people building such a set can take these extensions into account, and that any policy is either redundant or the kind of thing that would be set aside anyway in the situation.

Reply:
One more case: the `Haskell98` extension set presumably came into existence after some of its component parts were broadly acknowledged as design dead-ends (and thus `Legacy`). Introducing this set is nonetheless useful, as it helps people improve compatibility with older code. Here, any policy could have been set aside, but making a policy seems to not buy us anything. I certainly agree that we should be surprised if e.g. `GHC2035` includes any deprecated extensions when it's introduced in 2035, but how will a written policy actually help?